DOC/RF: add docstrings to `utils.py` and constants

aqw opened the pull request:
This PR is again eclectic, but is finally beginning to turn the corner to focus more on docstrings.

The PR includes:

- combine `assets.py` into `utils.py`
- retire the `Asset` type (which was just a `dict`)
- drop unused fsck tests
- consolidate `yaml_to_dict()` -> `get_asset_content()`
- rename `dict_to_yaml()` -> `dict_to_asset_yaml()`
- add docstrings to all functions in `utils.py` and constants in `consts.py`
- simplify `write_asset_file()`
- configured code coverage reports to ignore `TYPE_CHECKING` blocks

When documenting the constants, I intentionally did not wade into the issue of resolving the clear discrepancy of the names of PSEUDO_KEYS and RESERVED_KEYS vs how they are used. That is on my TODO list, but looks to be a complicated mess to unravel.

TobiasKadelka reviewed (commented):
Hey, this looks good to me :)

All my remarks are just little style and wording remarks, and the latter might be caused more by me than by the sentences itself.

aqw commented on onyo/lib/inventory.py:
```diff
@@ -347,8 +346,8 @@ def remove_asset(self, asset: Asset | Path) -> list[InventoryOperation]:
 
-    def move_asset(self, src: Path | Asset, dst: Path) -> list[InventoryOperation]:
-        if isinstance(src, Asset):
+    def move_asset(self, src: Path | dict, dst: Path) -> list[InventoryOperation]:
```
In general I agree, but I think it's worth explaining my approach here:

- I choose one file to work on docstrings
- As I improve the docstrings, I also make adjustments to the code itself (name, variables, etc)
- If those changes ripple out into other files, I make sure to make my changes as minimal as possible

If I didn't take the piece-by-piece approach, the chain-reaction of changes would result in a single PR becoming a massive rewrite of all of Onyo at once.

For many reasons, that is best avoided.

That is why I made the minimal changes possible to `inventory.py` (and others). My focus was on `assets.py`, `utils.py`, `consts.py`.

TobiasKadelka replied:
Thanks, that is fine with me.
Then I will be nitpicky about this when the time has come :)

aqw replied:
Nitpicking is definitely appreciated! Just on the files being focused on. :-)

aqw commented on onyo/lib/utils.py:
```diff
@@ -59,27 +82,78 @@ def get_asset_content(asset_file: Path) -> dict[str, float | int | str | Path]:
+    if invalid_yaml:
+        ui.error('The following files fail YAML validation:\n{}'.format(
+            '\n'.join(invalid_yaml)))
+
```
This is copy-paste from `assets.py`.

I will make the change for consistency, and push it.

aqw replied:
Done.

aqw commented on onyo/lib/utils.py:
```diff
@@ -59,27 +82,78 @@ def get_asset_content(asset_file: Path) -> dict[str, float | int | str | Path]:
+    duplicates = [a for a in asset_files if asset_names.count(a.name) > 1]
+    duplicates.sort(key=lambda x: x.name)
+
+    if duplicates:
```
Yeah; I noticed that too.

This was already uncovered. The point of this PR is not testing. I'm leaving testing for another effort.

Moving only one mountain at a time. ;-)

aqw commented on onyo/lib/utils.py:
```diff
@@ -59,27 +82,78 @@ def get_asset_content(asset_file: Path) -> dict[str, float | int | str | Path]:
+    Parameters
+    ----------
+    asset_files
+        A set of files to check the uniqueness of their file names.
```
Yeah, I rewrote this a couple times and didn't like them.

Reading it again, I also agree that it's unclear.

I've used your suggestion and will push it soon.

aqw replied:
Done.

aqw commented on onyo/lib/utils.py:
```diff
@@ -2,52 +2,75 @@
+    discarded.
 
-    For ease of use, accepts `None` (and returns it in that case).
+    For convenience, also accepts ``None`` and returns ``None`` in that case.
```
Good catch.

This stackoverflow comment has a nice overview that helped me think about this: https://stackoverflow.com/a/72805661

Basically, I like the idea of staying "imperative" whenever we can. However, to rewrite this in the imperative I think would be confusing (and imply that we're telling the user to do this, rather than describing what the function does)

> For convenience, also accepts ``None`` and returns ``None`` in that case.

to

> For convenience, also accept ``None`` and return ``None`` in that case.

That just feels weird.

The [pydocstyle project](https://www.pydocstyle.org/en/stable/error_codes.html) seems to only check for this on the first line. Which makes a lot of sense to me as a rule.

The short summary is imperative. But the extended summary is meant to explain a variety of things. Depending on the topic, the style will need to adapt.

aqw commented:
Thanks for the review!

I've addressed your comments. The only open one is about the voice. But I feel comfortable with the idea that only the short summary needs to be in the imperative.

TobiasKadelka commented:
Thanks a lot :)

I am fine with it all, and regarding the voice: It works obviously either way, it was just hard not to notice while reading trough it all.

From my side this PR can be merged as soon as the tests ran successfully :)

aqw commented:
Thanks!

Ideally, I will wait for Ben to review this. Specifically my changes to `yaml_to_dict()` and `write_asset_file()`.

Buuut we'll see if that responsibility survives the weekend. ;-)

Thanks again for your review!

aqw commented on onyo/lib/utils.py:
```diff
@@ -59,27 +82,77 @@ def get_asset_content(asset_file: Path) -> dict[str, float | int | str | Path]:
+        # TODO: use valid_yaml()
+        try:
+            YAML(typ='rt').load(asset)
+        except scanner.ScannerError:
```
I agree.

However... our current YAML parser (ruamel) isn't terribly good at providing useful error messages (see #587).

This is part of the motivation behind the (strictyaml](https://pypi.org/project/strictyaml/) project.

I will open an issue for that.

aqw replied:
Opened as #618

aqw commented:
I talked with Ben yesterday and got an in-person +1 to merge. :-)
